### PR TITLE
[BUG] Consider each site_package

### PR DESCRIPTION
**Pull request by BennettRand**
## Why are the changes needed?

`site.getsitepackages()` was returning
```
[
    '/home/username/project/python/.venv/lib/python3.10/site-packages',
    '/home/username/project/python/.venv/local/lib/python3.10/dist-packages',
    '/home/username/project/python/.venv/lib/python3/dist-packages',
    '/home/username/project/python/.venv/lib/python3.10/dist-packages'
]
```
which when run through `os.path.commonpath` is
```
'/home/username/project/python/.venv'
```
This makes it so that any `mod_file` compared to the site packages set, will never be a string in the site packages set, thus being included in the list of filed to be packaged for fast serialization.

## What changes were proposed in this pull request?

Compare the common path of the module file to each site package in the site package set to see if the mosule file should be skipped.

## How was this patch tested?

No tests added, all tests passed in Python 3.12.7

Manually inspected the file tree generated by a `pyflyte -v run --remote ...` command and the contents of the resulting `.tar.gz` file.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [ ] All commits are signed-off.

### CODE REVIEW

**Comment by wild-endeavor** [comment]
any chance you can think of a simple unit test for this?

**Review by thomasjpfan — COMMENTED** [review]
@BennettRand I think the simplest unit test is to monkeypatch `site.getsitepackages` and check that `list_imported_modules_as_files` does the right thing.

**Comment by BennettRand** [comment]
@thomasjpfan @wild-endeavor Just added it.
Should cover the general case and the corner case I ran into.

**Review by thomasjpfan — APPROVED** [review]
Thank you for the test! LGTM

**Review by pingsutw — APPROVED** [review]

File: tests/flytekit/unit/tools/test_script_mode.py [hunk 1]
```diff
@@ -237,3 +240,70 @@ def test_get_all_modules(tmp_path):
     # Workflow exists, so it is imported
     workflow_file.write_text(WORKFLOW_CONTENT)
     assert n_sys_modules + 1 == len(get_all_modules(os.fspath(source_dir), "my_workflows.main"))
+
+
+def test_list_imported_modules_as_files():
+
+    source_path = "/home/username/project"
```
Comment (thomasjpfan) [inline tests/flytekit/unit/tools/test_script_mode.py:247]: For windows CI to pass, I think we'll either need to use `os.path.join` or `pathlib.Path` to define the paths.

(Windows uses `\` instead of `/`)

**Comment by BennettRand** [comment]
@thomasjpfan Yeah, I believe my unit test is showing that list_imported_modules_as_files is not behaving as expected in Windows. I'm working on launching a Windows VM to debug it properly.

**Comment by BennettRand** [comment]
`os.path.commonpath` was raising a `ValueError` mid-generator, leading it to hit the `pass` statement in the `except` block and falsely including some modules.

File: flytekit/tools/script_mode.py [hunk 2]
```diff
@@ -183,6 +183,14 @@ def list_all_files(source_path: str, deref_symlinks, ignore_group: Optional[Igno
     return all_files
 
 
+def __drive_safe_commonpath(paths):
```
Comment (thomasjpfan) [inline flytekit/tools/script_mode.py:186]: Can you use one underscore here?

```suggestion
def _drive_safe_commonpath(paths):
```

File: flytekit/tools/script_mode.py [hunk 3]
```diff
@@ -231,6 +235,12 @@ def list_imported_modules_as_files(source_path: str, modules: List[ModuleType])
             # in the site-packages or bin directory.
             pass
 
+        if any(
+            (__drive_safe_commonpath([site_package, mod_file]) == site_package for site_package in site_packages_set)
```
Comment (thomasjpfan) [inline flytekit/tools/script_mode.py:239]: Given the access pattern, I think we can remove `site_packages_set` and use `site_packages` directly:

```suggestion
            (__drive_safe_commonpath([site_package, mod_file]) == site_package for site_package in site_packages)
```

File: tests/flytekit/unit/tools/test_script_mode.py [hunk 4]
```diff
@@ -237,3 +242,71 @@ def test_get_all_modules(tmp_path):
...
     workflow_file.write_text(WORKFLOW_CONTENT)
     assert n_sys_modules + 1 == len(get_all_modules(os.fspath(source_dir), "my_workflows.main"))
+
+
+def test_list_imported_modules_as_files():
+
+    bin_directory = Path(os.path.dirname(sys.executable))
+    flytekit_root = Path(os.path.dirname(flytekit.__file__))
+    root_dir = Path("C:\\") if platform.system() == "Windows" else Path("/")
+    source_path = root_dir / "home" / "username" / "project"
```
Comment (thomasjpfan) [inline tests/flytekit/unit/tools/test_script_mode.py:252]: To keep this simple, can we have the `source_path` be `Path.cwd()`?

File: tests/flytekit/unit/tools/test_script_mode.py [hunk 5]
```diff
@@ -237,3 +242,71 @@ def test_get_all_modules(tmp_path):
...
+        str(Path(site_packages[1]) / "package" / "module_2.py"),
+        str(Path(site_packages[2]) / "package" / "module_3.py"),
+        str(Path(site_packages[3]) / "package" / "module_4.py"),
+    ]
+
+    modules = local_modules + flyte_modules + bin_modules + site_modules
+    module_paths = local_module_paths + flyte_module_paths + bin_module_paths + site_module_paths
+
+    for m, p in zip(modules, module_paths):
+        m.__file__ = p
```
Comment (thomasjpfan) [inline tests/flytekit/unit/tools/test_script_mode.py:307]: Depending on two list being aligned feels a little brittle. Can have the modules be a list of tuples? Something like this:

```python
local_modules = [
    (ModuleType("local_module_1"), str(source_path / "package_a" / "module_1.py")),
    ...
]

all_modules = local_modules + flyte_modules + ...

for mod, mod_path in all_modules:
    mod.__file__  = mod_path

# to be passed in later
modules = [mod for mod, _ in all_modules]

with mock.patch("site.getsitepackages", new=lambda: site_packages):
    file_list = list_imported_modules_as_files(str(source_path), modules)

local_module_paths = [path for _, path in local_modules]
assert sorted(file_list) == sorted(local_module_paths)
```

**Comment by thomasjpfan** [comment]
@BennettRand Can you sync with `master` to see if it fixes the CI issues?

**Comment by wild-endeavor** [comment]
closing for now - @BennettRand could you confirm release 1.14.0 fixes the issue you were seeing?
